### PR TITLE
Upstream the Unicode database table generator and update the tables to 15

### PR DESCRIPTION
**Pull request by rikkimax**
Right let's see how much this breaks.

### CODE REVIEW

**Comment by burner** [comment]
Considering the size alone of this PR. I would welcome a somewhat long description of what this PR tries to do.
I think I have an idea (update the unicode in phobos), but I might just be overconfident.

**Comment by rikkimax** [comment]
> Considering the size alone of this PR. I would welcome a somewhat long description of what this PR tries to do. I think I have an idea (update the unicode in phobos), but I might just be overconfident.

The goal is to get the Unicode table generator into Phobos where it belongs.

From what I can tell, it has never been updated and could be ~8 releases out of date.

Unfortunately, people went messing with the tables since then, and not all changes got committed to the generator so it's not quite as simple as just adding it.

The quality of the generator doesn't matter, nor do the tables themselves. As long as it passes the test suite we are good (Dmitry attempted to update the tables to 14 already, and this approach was the outcome of that).

File: std/internal/unicode_table_generator.d [hunk 1]
```diff
@@ -0,0 +1,1247 @@
...
+            sink.writef("//%d bytes\nenum %sTrieEntries = TrieEntry!(%s",
+                trie.bytes, name, Unqual!(typeof(T.init[0])).stringof);
+            foreach (lvl; Params[0..$])
+                sink.writef(", %d", lvl);
+            sink.write(")(");
+            trie.store(sink.lockingTextWriter());
+            sink.writeln(");");
+        };
+    }
+}
```
Comment (ljmf00) [inline std/internal/unicode_table_generator.d:1247]: missing new line

**Comment by rikkimax** [comment]
@ljmf00 I didn't develop the generator, that was Dmitry Olshansky ~2013. Its quality does not matter as long as it works. If we do not get the generator into Phobos ``std.uni`` is effectively dead.

Right now I want to change as little as possible, it needs to work, beyond that it's out of scope of the PR.

**Comment by rikkimax** [comment]
Okay looks like the testsuite has got the brilliant idea that it needs to run the generator.

> std.exception.ErrnoException@std/stdio.d(547): Cannot open file `ucd-15/Blocks.txt' in mode `rb' (No such file or directory)

But otherwise its looking like its passing :D

**Comment by ibuclaw** [comment]
> Okay looks like the testsuite has got the brilliant idea that it needs to run the generator.
> 
> > std.exception.ErrnoException@std/stdio.d(547): Cannot open file `ucd-15/Blocks.txt' in mode `rb' (No such file or directory)
> 
> But otherwise its looking like its passing :D

The generator should sit outside of the std source tree, such as in a `tools` directory.

File: posix.mak [hunk 2]
```diff
@@ -591,7 +591,8 @@ style_lint_shellcmds:
 	grep -nr '[[:blank:]]$$' $$(find etc std -name '*.d'); test $$? -eq 1
 
 	@echo "Enforce whitespace before opening parenthesis"
-	grep -nrE "\<(for|foreach|foreach_reverse|if|while|switch|catch|version)\(" $$(find etc std -name '*.d') ; test $$? -eq 1
+    # do not do whitespace checks on generated unicode tables
+	grep -nrE "\<(for|foreach|foreach_reverse|if|while|switch|catch|version)\(" $$(find etc std -name '*.d' -not -wholename 'std/internal/unicode_*.d') ; test $$? -eq 1
```
Comment (ibuclaw) [inline posix.mak:595]: It shouldn't be too difficult to make the generator emit compliant code (it's just a space before `(`).

If only the generator itself falls foul of code style, then having it in another directory would remove the need for this.

File: posix.mak [hunk 3]
```diff
@@ -603,7 +604,8 @@ style_lint_shellcmds:
 	grep -nr 'import std.algorithm : ' $$(find etc std -name '*.d') ; test $$? -eq 1
 
 	@echo "Enforce Allman style"
-	grep -nrE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .*{$$' $$(find etc std -name '*.d'); test $$? -eq 1
+    # do not do whitespace checks on generated unicode tables
+	grep -nrE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .*{$$' $$(find etc std -name '*.d' -not -wholename 'std/internal/unicode_*.d'); test $$? -eq 1
```
Comment (ibuclaw) [inline posix.mak:608]: Ditto this (add a newline between `) {`).

File: tools/unicode_table_generator.d [hunk 4]
```diff
@@ -0,0 +1,1296 @@
...
+        writeTableOfSets(baseSink, "uniProps", general);
+        writeTableOfSets(baseSink, "blocks", blocks);
+        writeTableOfSets(baseSink, "scripts", scripts);
+        writeTableOfSets(baseSink, "hangul", hangul);
+        writeFunctions(baseSink);
+    }
+
+    static void trieProlog(File file)
+    {
+        file.writefln("\nstatic if (size_t.sizeof == %d) {", size_t.sizeof);
```
Comment (ibuclaw) [inline tools/unicode_table_generator.d:251]:
```suggestion
        file.writefln("\nstatic if (size_t.sizeof == %d)\n{", size_t.sizeof);
```

File: posix.mak (hunk 2 above)
Comment (rikkimax) [reply posix.mak:595]: Unfortunately, some of the generated files (such as long lines) are indeed a bit hard to fix.

So I'm opting to just disable the checks, as my focus is on the Unicode side of things, not making style checkers happy ;)

File: posix.mak (hunk 3 above)
Comment (rikkimax) [reply posix.mak:608]: Done locally

**Review by ibuclaw — COMMENTED** [review]
Can't see anywhere else that generated code would fall foul of the crude style checks.

File: posix.mak (hunk 2 above)
Comment (ibuclaw) [reply posix.mak:595]: For dscanner, disabling checks such as line length and others is completely fine. For these crude tests, I can only see one place in the generator that would trigger this.

**Comment by rikkimax** [comment]
Buildkite is failing due to a bug in mir-algorithm, so it's safe to assume as of right now, that this is green!

File: posix.mak [hunk 5]
```diff
@@ -591,6 +591,7 @@ style_lint_shellcmds:
 	grep -nr '[[:blank:]]$$' $$(find etc std -name '*.d'); test $$? -eq 1
 
 	@echo "Enforce whitespace before opening parenthesis"
+    # do not do whitespace checks on generated unicode tables
```
Comment (ibuclaw) [inline posix.mak:594]:
```suggestion
```

File: posix.mak [hunk 6]
```diff
@@ -603,7 +604,9 @@ style_lint_shellcmds:
 	grep -nr 'import std.algorithm : ' $$(find etc std -name '*.d') ; test $$? -eq 1
 
 	@echo "Enforce Allman style"
-	grep -nrE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .*{$$' $$(find etc std -name '*.d'); test $$? -eq 1
+    # do not do whitespace checks on generated unicode tables
+	grep -nrE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .* {$$' \
+    $$(find etc std -name '*.d' -not -wholename 'std/internal/unicode_*.d'); test $$? -eq 1
```
Comment (ibuclaw) [inline posix.mak:608]:
```suggestion
	grep -nrE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .*{$$' $$(find etc std -name '*.d'); test $$? -eq 1
```
Pretty sure this is no longer needed. :-)

Comment (rikkimax) [reply posix.mak]: Let's find out!

**Comment by ibuclaw** [comment]
> Buildkite is failing due to a bug in mir-algorithm, so it's safe to assume as of right now, that this is green!

It's only failing with this PR though, so perhaps you've indirectly affected something?

https://github.com/libmir/mir-algorithm/blob/aa1e914663999eb94d89e84a30c645ea94def179/source/mir/format.d#L909-L918

**Comment by ibuclaw** [comment]
> > Buildkite is failing due to a bug in mir-algorithm, so it's safe to assume as of right now, that this is green!
> 
> It's only failing with this PR though, so perhaps you've indirectly affected something?
> 
> https://github.com/libmir/mir-algorithm/blob/aa1e914663999eb94d89e84a30c645ea94def179/source/mir/format.d#L909-L918

Answer, yes: see Mir's implementation of `print`

https://github.com/libmir/mir-algorithm/blob/e0acce4313a26385d2e461f58ec5a1a7a6a3acc1/source/mir/format.d#L882-L906

**Comment by rikkimax** [comment]
> > > Buildkite is failing due to a bug in mir-algorithm, so it's safe to assume as of right now, that this is green!
> > 
> > 
> > It's only failing with this PR though, so perhaps you've indirectly affected something?
> > https://github.com/libmir/mir-algorithm/blob/aa1e914663999eb94d89e84a30c645ea94def179/source/mir/format.d#L909-L918
> 
> Answer, yes: see Mir's implementation of `print`
> 
> https://github.com/libmir/mir-algorithm/blob/e0acce4313a26385d2e461f58ec5a1a7a6a3acc1/source/mir/format.d#L882-L906
> 
> So I'd consider you to be the owner of fixing Mir if you're going to break downstream. :-)

Yeah it's calling into isGraphical, and the character in question wasn't allocated before 7 so it couldn't be graphical in previous tables.

Should be an easy fix, remove one character from the unittest methodology effectively.

File: changelog/unicode_table_generator.dd [hunk 7]
```diff
@@ -0,0 +1,4 @@
+Unicode table generator is now in Phobos, tables are updated to version 15.
+
+It is likely that this change will result in breakage in code and program usage.
+This is due to a number of factors, the tables being updated so significantly and the table generator not having all its changes commited throughout the years.
```
Comment (ibuclaw) [inline changelog/unicode_table_generator.dd:4]: Does your editor support adding newlines to the end of files? :-)

Comment (rikkimax) [reply changelog/unicode_table_generator.dd]: You know, it's the oddest thing, Nodepad++ has literally every other feature but just not that one!

**Comment by rikkimax** [comment]
Looks like I'll have to remove myself as a code owner since I don't have write permissions

**Comment by RazvanN7** [comment]
cc @atilaneves . Any thoughts on this?

**Comment by atilaneves** [comment]
I agree with @ibuclaw in that the generator should probably be in tools but it seems like it is already?

**Comment by rikkimax** [comment]
> I agree with @ibuclaw in that the generator should probably be in tools but it seems like it is already?

It wasn't originally, but I applied that change since it made sense.

**Comment by rikkimax** [comment]
> std/traits.d(8779): Error: static assert:  `1LU == 2LU` is false

what the?

I see in another PR has this exact issue for the same failed tests. So not my fault, phew.

**Comment by RazvanN7** [comment]
Rebasing to latest master will fix this.

**Comment by rikkimax** [comment]
Oh dangit I shouldn't have let SourceTree merge branches like that. Oh well, I'll let CI run and confirm the fix for prior error.

File: tools/unicode_table_generator.d [hunk 8]
```diff
@@ -0,0 +1,1303 @@
...
+
+enum HeaderComment = `//Written in the D programming language
+/**
+ * License: $(HTTP boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ *
+ * Authors: Dmitry Olshansky
+ *
+ */
+// Automatically generated from Unicode Character Database files, DO NOT EDIT
+// https://github.com/dlang/phobos/blob/master/tools/unicode_table_generator.d
```
Comment (ibuclaw) [inline tools/unicode_table_generator.d:169]:
```suggestion
// !!! DO NOT EDIT !!!
// This module is automatically generated from Unicode Character Database files
// See: https://github.com/dlang/phobos/blob/master/tools/unicode_table_generator.d
```

Comment (rikkimax) [reply tools/unicode_table_generator.d:169]: Considering people have edited these files in the past?

Sounds like a plan! lol

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@atilaneves @ibuclaw any other objections or is this good to go?

**Pull request merged**